ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java

```diff
@@ -388,6 +403,61 @@ private void checkAndMaybeComplete() {
     }
   }
 
+  /**
+   * Try to replicate the blob and then retry the TtlUpdate Request
+   * @return true if it's under retry. return false if cannot do retry or retry is finished.
+   */
+  private boolean retryWithReplicateBlob() {
+    // the conditions to trigger ReplicateBlob operation:
+    // 1. at least one replica returned successful status
+    // 2. and at least one replica returned NOT_FOUND status.
+    if (!operationTracker.hasNotFound() || operationTracker.getSuccessReplica().size() == 0) {
+      return false;
+    }
+
+    List<ReplicaId> successfulReplica = operationTracker.getSuccessReplica();
+    // it's the first time we hit the 503
+    if (replicateBlobCallback == null) {
+      DataNodeId sourceDataNode = successfulReplica.get(0).getDataNodeId();
+      replicateBlobCallback = new ReplicateBlobCallback(blobId, sourceDataNode);
+
+      nonBlockingRouter.replicateBlob(blobId.getID(), this.getClass().getSimpleName(), sourceDataNode,
+          replicateBlobCallback);
+      LOGGER.trace("Start the on-demand replication {} {}.", replicateBlobCallback.getBlobId(),
+          replicateBlobCallback.getSourceDataNode());
+      return true;
+    }
+
+    ReplicateBlobCallback.RetryState state = replicateBlobCallback.getState();
+    switch (state) {
+      case REPLICATING_BLOB:
+        LOGGER.trace("Running on-demand replication {} {}.", replicateBlobCallback.getBlobId(),
+            replicateBlobCallback.getSourceDataNode());
+        return true;
+      case REPLICATION_DONE:
+        if (replicateBlobCallback.getException() != null) {
+          LOGGER.error("on-demand replication {} {} failed.", replicateBlobCallback.getBlobId(),
+              replicateBlobCallback.getSourceDataNode(), replicateBlobCallback.getException());
+          return false;
+        } else {
+          LOGGER.info("On-demand replication {} {} is successful.", replicateBlobCallback.getBlobId(),
+              replicateBlobCallback.getSourceDataNode());
+          replicateBlobCallback.setState(ReplicateBlobCallback.RetryState.RETRYING);
+
+          // retry the TtlUpdate
+          operationException.set(null);
+          operationTracker =
+              new SimpleOperationTracker(routerConfig, RouterOperation.TtlUpdateOperation, blobId.getPartition(),
+                  originatingDcName, false, routerMetrics, blobId);
+          return true;
+        }
+      case RETRYING:
+      default:
+        // retry is done but still failed. Nothing else we can do.
```

Comment:
Are we suppose to through exception when we hit this point? We only support three enum type right?